tapiriik/services/PolarFlow/polarflow.py

```diff
@@ -0,0 +1,270 @@
+
+    def _create_activity(self, activity_data):
+        activity = UploadedActivity()
+
+        activity.Stationary = not activity_data["has-route"]
+        if "detailed-sport-info" in activity_data and activity_data["detailed-sport-info"] in self._reverse_activity_type_mappings:
+            activity.Type = self._reverse_activity_type_mappings[activity_data["detailed-sport-info"]]
+        else:
+            activity.Type = ActivityType.Other
+
+        activity.StartTime = pytz.utc.localize(isodate.parse_datetime(activity_data["start-time"]))
+        activity.EndTime = activity.StartTime + isodate.parse_duration(activity_data["duration"])
+
+        distance = activity_data["distance"] if "distance" in activity_data else None
+        activity.Stats.Distance = ActivityStatistic(ActivityStatisticUnit.Meters, value=float(distance) if distance else None)
+        hr_data = activity_data["heart-rate"] if "heart-rate" in activity_data else None
+        avg_hr = hr_data["average"] if "average" in hr_data else None
+        max_hr = hr_data["maximum"] if "maximum" in hr_data else None
+        activity.Stats.HR.update(ActivityStatistic(ActivityStatisticUnit.BeatsPerMinute, avg=float(avg_hr) if avg_hr else None, max=float(max_hr) if max_hr else None))
+        calories = activity_data["calories"] if "calories" in activity_data else None
+        activity.Stats.Energy = ActivityStatistic(ActivityStatisticUnit.Kilocalories, value=int(calories) if calories else None)
+
+        activity.ServiceData = {"ActivityID": activity_data["id"]}
+
+        logger.debug("\tActivity s/t {}: {}".format(activity.StartTime, activity.Type))
+
+        activity.CalculateUID()
+        return activity
+
+    def DownloadActivity(self, serviceRecord, activity):
+        # NOTE tcx have to be gzipped but it actually doesn't
+        # https://www.polar.com/accesslink-api/?python#get-tcx
+        #tcx_data_raw = requests.get(activity_link + "/tcx", headers=self._api_headers(serviceRecord))
+        #tcx_data = gzip.GzipFile(fileobj=StringIO(tcx_data_raw)).read()
+        tcx_url = serviceRecord.ServiceData["Transaction-uri"] + "/exercises/{}/tcx".format(activity.ServiceData["ActivityID"])
+        response = requests.get(tcx_url, headers=self._api_headers(serviceRecord, {"Accept": "application/vnd.garmin.tcx+xml"}))
+        if response.status_code == 404:
+            # Transaction was disbanded, all data linked to it will be returned in next transaction
+            raise APIException("Transaction disbanded", block=True, user_exception=UserException(UserExceptionType.DownloadError))
+        try:
+            tcx_data = response.text
+            activity = TCXIO.Parse(tcx_data.encode('utf-8'), activity)
+            #TODO: uncomment when interchange will be ready
+            #activity.SourceFile = SourceFile(tcx_data, ActivityFileType.TCX)
+        except lxml.etree.XMLSyntaxError:
+            logger.debug("Cannot recieve training tcx at url: {}".format(tcx_url))
```

Comment:
Would want to raise an APIException here, otherwise the sync logic will think our empty activity is valid and attempt to upload it to other services.

Reply:
Sounds reasonable for me too.